Error moving a single custom button

### Steps to reproduce

1. Fresh Pronterface run, i.e. no `~/.config/Printrun/pronsolerc` file around.
2. Create a custom button.
3. Right click on new button and hit "Move right >>".

The following is printed to the Pronteface log (interestingly, nothing on the terminal):

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/wx/core.py", line 3427, in <lambda>
    lambda event: event.callable(*event.args, **event.kw) )
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/test/printrun/pronterface.py", line 2155, in cbuttons_reload
    self.cbuttonssizer.Add(b, flag = wx.EXPAND)
wx._core.wxAssertionError: C++ assertion "CheckExpectedParentIs(w, m_containingWindow)" failed at ./src/common/sizer.cpp(850) in DoInsert(): Windows managed by the sizer associated with the given window must have this window as parent, otherwise they will not be repositioned correctly.

Please use the window wxPanel@0x3e9e910 ("panel") with which this sizer is associated, as the parent when creating the window wxStaticText@0x3feb500 ("staticText") managed by it.
```

--- a/printrun/gui/widgets.py
+++ b/printrun/gui/widgets.py
@@ -15,9 +15,10 @@
 
 import wx
 import re
+import platform
 
 def getSpace(a):
-    #this method is used to dehardcode and unify borders and gaps of the interface
+    # This method is used to de-hardcode and unify the borders and gaps of the interface.
     match a:
         case 'major': #e.g. outer border of dialog boxes
             return 12
@@ -25,6 +26,15 @@ def getSpace(a):
             return 8
         case 'mini':
             return 4
+        case 'stddlg':
+            # Differentiation is necessary because wxPyhton behaves slightly differently on different systems.
+            platformname = platform.system()
+            if platformname == 'Windows':
+                return 8
+            elif platformname == 'Darwin':
+                return 4
+            else:
+                return 4 # Not sure yet which value should be used for linux systems
         case 'none':
             return 0
 
@@ -73,7 +83,7 @@ class MacroEditor(wx.Dialog):
         btnsizer.AddButton(self.savebtn)
         btnsizer.AddButton(self.cancelbtn)
         btnsizer.Realize()
-        topsizer.Add(btnsizer, 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('minor'))
+        topsizer.Add(btnsizer, 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('stddlg'))
         self.SetSizer(topsizer)
         self.SetMinClientSize((230,150))
         topsizer.Fit(self)
@@ -88,7 +98,7 @@ class MacroEditor(wx.Dialog):
         selection = self.e.GetStringSelection()
         if selection:
             self.finddata.SetFindString(selection)
-        self.finddialog = wx.FindReplaceDialog(self.e, self.finddata, _("Find and replace"), wx.FR_NOWHOLEWORD) #wx.FR_REPLACEDIALOG
+        self.finddialog = wx.FindReplaceDialog(self.e, self.finddata, _("Find..."), wx.FR_NOWHOLEWORD) #wx.FR_REPLACEDIALOG
         # TODO: Setup ReplaceDialog, Setup WholeWord Search, deactivated for now...
         self.finddialog.Show()
 
@@ -240,7 +250,7 @@ class PronterOptionsDialog(wx.Dialog):
         topsizer = wx.BoxSizer(wx.VERTICAL)
         topsizer.Add(notebook, 1, wx.EXPAND | wx.ALL, getSpace('minor'))
         topsizer.Add(wx.StaticLine(self, -1, style = wx.LI_HORIZONTAL), 0, wx.EXPAND)
-        topsizer.Add(self.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('minor'))
+        topsizer.Add(self.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('stddlg'))
         self.SetSizer(topsizer)
         self.Fit()
         self.CentreOnParent()
@@ -268,19 +278,14 @@ class ButtonEdit(wx.Dialog):
         wx.Dialog.__init__(self, None, title = _("Custom button"),
                            style = wx.DEFAULT_DIALOG_STYLE)
         self.pronterface = pronterface
-        # All widgets except ok/cancel are grouped into a static box for aesthetic and ux reasons
         panel = wx.Panel(self)
-        #panelsizer = wx.BoxSizer
-        #self.sbox = wx.StaticBox(self)
-        #sboxsizer = wx.StaticBoxSizer(self.sbox, wx.VERTICAL)
         grid = wx.FlexGridSizer(rows = 0, cols = 2, hgap = getSpace('minor'), vgap = getSpace('minor'))
         grid.AddGrowableCol(1, 1)
         ## Title of the button
         grid.Add(wx.StaticText(panel, -1, _("Button title:")), 0, wx.ALIGN_CENTER_VERTICAL | wx.ALIGN_RIGHT)
         self.name = wx.TextCtrl(panel, -1, "")
-        min_size = self.name.GetTextExtent('Default Long Button Name')
-        min_size.width = 260 #overwrite width to make the dialog a bit wider
-        self.name.SetMinSize(wx.Size(min_size.width, -1))
+        dlg_size = 260
+        self.name.SetMinSize(wx.Size(dlg_size, -1))
         grid.Add(self.name, 1, wx.EXPAND)
         ## Colour of the button
         grid.Add(wx.StaticText(panel, -1, _("Colour:")), 0, wx.ALIGN_CENTER_VERTICAL | wx.ALIGN_RIGHT)
@@ -310,13 +315,11 @@ class ButtonEdit(wx.Dialog):
         self.macrobtn.Bind(wx.EVT_BUTTON, self.macrob_handler)
         commandsizer.Add(self.macrobtn, 0, wx.ALIGN_CENTER_VERTICAL | wx.LEFT, getSpace('minor'))
         grid.Add(commandsizer, 1, wx.EXPAND)
-        ## Put it in a box and add ok/cancel buttons
-        #panel.Add(grid, 1, wx.EXPAND | wx.ALL, getSpace('mini'))
         panel.SetSizer(grid)
         topsizer = wx.BoxSizer(wx.VERTICAL)
         topsizer.Add(panel, 0, wx.EXPAND | wx.ALL, getSpace('major'))
         topsizer.Add(wx.StaticLine(self, -1, style = wx.LI_HORIZONTAL), 0, wx.EXPAND)
-        topsizer.Add(self.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('minor'))
+        topsizer.Add(self.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('stddlg'))
         self.SetSizer(topsizer)
         topsizer.Fit(self)
         self.CentreOnParent()

--- a/printrun/gui/widgets.py
+++ b/printrun/gui/widgets.py
@@ -16,64 +16,124 @@
 import wx
 import re
 
+def getSpace(a):
+    #this method is used to dehardcode and unify borders and gaps of the interface
+    match a:
+        case 'major': #e.g. outer border of dialog boxes
+            return 12
+        case 'minor': #e.g. border of inner elements
+            return 8
+        case 'mini':
+            return 4
+        case 'none':
+            return 0
+
 class MacroEditor(wx.Dialog):
     """Really simple editor to edit macro definitions"""
 
     def __init__(self, macro_name, definition, callback, gcode = False):
         self.indent_chars = "  "
-        title = "  macro %s"
+        title = "  Macro %s"
         if gcode:
             title = "  %s"
         self.gcode = gcode
         wx.Dialog.__init__(self, None, title = title % macro_name,
                            style = wx.DEFAULT_DIALOG_STYLE | wx.RESIZE_BORDER)
         self.callback = callback
-        self.panel = wx.Panel(self, -1)
+        panel = wx.Panel(self)
+        panelsizer = wx.BoxSizer(wx.VERTICAL)
         titlesizer = wx.BoxSizer(wx.HORIZONTAL)
-        self.titletext = wx.StaticText(self.panel, -1, "              _")  # title%macro_name)
-        titlesizer.Add(self.titletext, 1)
-        self.findb = wx.Button(self.panel, -1, _("Find"), style = wx.BU_EXACTFIT)  # New button for "Find" (Jezmy)
-        self.findb.Bind(wx.EVT_BUTTON, self.find)
-        self.okb = wx.Button(self.panel, -1, _("Save"), style = wx.BU_EXACTFIT)
-        self.okb.Bind(wx.EVT_BUTTON, self.save)
+        self.titletext = wx.StaticText(panel, -1, "              _")  # title%macro_name)
+        titlesizer.Add(self.titletext, 1, wx.ALIGN_CENTER_VERTICAL)
+        self.findbtn = wx.Button(panel, -1, _(" Find "), style = wx.BU_EXACTFIT)  # New button for "Find" (Jezmy)
+        self.findbtn.Bind(wx.EVT_BUTTON, self.on_find)
+        self.Bind(wx.EVT_FIND, self.on_find_find)
+        self.Bind(wx.EVT_FIND_NEXT, self.on_find_find)
+        self.Bind(wx.EVT_FIND_CLOSE, self.on_find_cancel)
         self.Bind(wx.EVT_CLOSE, self.close)
-        titlesizer.Add(self.findb)
-        titlesizer.Add(self.okb)
-        self.cancelb = wx.Button(self.panel, -1, _("Cancel"), style = wx.BU_EXACTFIT)
-        self.cancelb.Bind(wx.EVT_BUTTON, self.close)
-        titlesizer.Add(self.cancelb)
-        topsizer = wx.BoxSizer(wx.VERTICAL)
-        topsizer.Add(titlesizer, 0, wx.EXPAND)
-        self.e = wx.TextCtrl(self.panel, style = wx.HSCROLL | wx.TE_MULTILINE | wx.TE_RICH2, size = (400, 400))
+        titlesizer.Add(self.findbtn, 0, wx.ALIGN_CENTER_VERTICAL)
+        panelsizer.Add(titlesizer, 0, wx.EXPAND | wx.ALL, getSpace('minor'))
+        self.e = wx.TextCtrl(panel, style = wx.HSCROLL | wx.TE_MULTILINE | wx.TE_RICH2, size = (400, 400)) #style = wx.HSCROLL | 
         if not self.gcode:
             self.e.SetValue(self.unindent(definition))
         else:
             self.e.SetValue("\n".join(definition))
-        topsizer.Add(self.e, 1, wx.ALL | wx.EXPAND)
-        self.panel.SetSizer(topsizer)
-        topsizer.Layout()
+        panelsizer.Add(self.e, 1, wx.EXPAND)
+        panel.SetSizer(panelsizer)
+        topsizer = wx.BoxSizer(wx.VERTICAL)
+        topsizer.Add(panel, 1, wx.EXPAND | wx.ALL, getSpace('none'))
+        # No StaticLine in this case bc the TextCtrl acts as a divider
+        #topsizer.Add(wx.StaticLine(self, -1, style = wx.LI_HORIZONTAL), 0, wx.EXPAND)
+        btnsizer = wx.StdDialogButtonSizer()
+        self.savebtn = wx.Button(self, wx.ID_SAVE)
+        self.savebtn.SetDefault()
+        self.savebtn.Bind(wx.EVT_BUTTON, self.save)
+        self.cancelbtn = wx.Button(self, wx.ID_CANCEL)
+        self.cancelbtn.Bind(wx.EVT_BUTTON, self.close)
+        btnsizer.AddButton(self.savebtn)
+        btnsizer.AddButton(self.cancelbtn)
+        btnsizer.Realize()
+        topsizer.Add(btnsizer, 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('minor'))
+        self.SetSizer(topsizer)
+        self.SetMinClientSize((230,150))
         topsizer.Fit(self)
+        self.CentreOnParent()
         self.Show()
         self.e.SetFocus()
 
-    def find(self, ev):
+    def on_find(self, ev):
         # Ask user what to look for, find it and point at it ...  (Jezmy)
-        S = self.e.GetStringSelection()
-        if not S:
-            S = "Z"
-        FindValue = wx.GetTextFromUser('Please enter a search string:', caption = "Search", default_value = S, parent = None)
-        somecode = self.e.GetValue()
-        position = somecode.find(FindValue, self.e.GetInsertionPoint())
-        if position == -1:
-            self.titletext.SetLabel(_("Not Found!"))
-        else:
-            self.titletext.SetLabel(str(position))
+        self.findbtn.Disable()
+        self.finddata = wx.FindReplaceData(wx.FR_DOWN)   # initializes and holds search parameters #wx.FR_DOWN
+        selection = self.e.GetStringSelection()
+        if selection:
+            self.finddata.SetFindString(selection)
+        self.finddialog = wx.FindReplaceDialog(self.e, self.finddata, _("Find and replace"), wx.FR_NOWHOLEWORD) #wx.FR_REPLACEDIALOG
+        # TODO: Setup ReplaceDialog, Setup WholeWord Search, deactivated for now...
+        self.finddialog.Show()
 
-            # ananswer = wx.MessageBox(str(numLines)+" Lines detected in file\n"+str(position), "OK")
-            self.e.SetFocus()
-            self.e.SetInsertionPoint(position)
-            self.e.SetSelection(position, position + len(FindValue))
-            self.e.ShowPosition(position)
+    def on_find_cancel(self, ev):
+        self.findbtn.Enable()
+        self.titletext.SetLabel("              _")
+        self.finddialog.Destroy()
+    
+    def on_find_find(self, ev):
+        findstring = self.finddata.GetFindString()
+        macrocode = self.e.GetValue()
+        
+        if self.e.GetStringSelection().lower() == findstring.lower():
+            # If the desired string is already selected, change the position to jump to the next one.
+            if self.finddata.GetFlags() % 2 == 1:
+                self.e.SetInsertionPoint(self.e.GetInsertionPoint() + len(findstring))
+            else:
+                self.e.SetInsertionPoint(self.e.GetInsertionPoint() - len(findstring))
+        
+        if int(self.finddata.GetFlags() / 4) != 1:
+            # When search is not case-sensitve, convert the whole string to lowercase
+            findstring = findstring.casefold()
+            macrocode = macrocode.casefold()
+
+        # The user can choose to search upwards or downwards
+        if self.finddata.GetFlags() % 2 == 1:
+            stringpos = macrocode.find(findstring, self.e.GetInsertionPoint())
+        else:
+            stringpos = macrocode.rfind(findstring, 0, self.e.GetInsertionPoint())
+        
+        if stringpos == -1 and self.finddata.GetFlags() % 2 == 1:
+            self.titletext.SetLabel(_("End of macro, jumped to first line"))
+            stringpos = 0 #jump to the beginning
+            self.e.SetInsertionPoint(stringpos)
+            self.e.ShowPosition(stringpos)
+        elif stringpos == -1 and self.finddata.GetFlags() % 2 == 0:
+            self.titletext.SetLabel(_("Begin of macro, jumped to last line"))
+            stringpos = self.e.GetLastPosition() #jump to the end
+            self.e.SetInsertionPoint(stringpos)
+            self.e.ShowPosition(stringpos)
+        else:
+            # TODO: Implement a Not Found state when no single match was found
+            self.titletext.SetLabel(_("Found!"))
+            self.e.SetSelection(stringpos, stringpos + len(findstring))
+            self.e.ShowPosition(stringpos)
 
     def ShowMessage(self, ev, message):
         dlg = wx.MessageDialog(self, message,
@@ -126,11 +186,8 @@ class PronterOptionsDialog(wx.Dialog):
     """Options editor"""
     def __init__(self, pronterface):
         wx.Dialog.__init__(self, parent = None, title = _("Edit settings"),
-                           size = (400, 500), style = wx.DEFAULT_DIALOG_STYLE)
-        panel = wx.Panel(self)
-        header = wx.StaticBox(panel, label = _("Settings"))
-        sbox = wx.StaticBoxSizer(header, wx.VERTICAL)
-        self.notebook = notebook = wx.Notebook(panel)
+                           size = wx.DefaultSize, style = wx.DEFAULT_DIALOG_STYLE)
+        self.notebook = notebook = wx.Notebook(self)
         all_settings = pronterface.settings._all_settings()
         group_list = []
         groups = {}
@@ -143,27 +200,33 @@ class PronterOptionsDialog(wx.Dialog):
                 groups[setting.group] = []
             groups[setting.group].append(setting)
         for group in group_list:
-            grouppanel = wx.Panel(notebook, -1)
+            grouppanel = wx.ScrolledWindow(notebook, -1, style = wx.VSCROLL)
+            # Setting the scrollrate based on the systemfont
+            fontsize = wx.SystemSettings.GetFont(wx.SYS_SYSTEM_FONT).GetPixelSize()
+            grouppanel.SetScrollRate(fontsize.x, fontsize.y)
             notebook.AddPage(grouppanel, SETTINGS_GROUPS[group])
             settings = groups[group]
-            grid = wx.GridBagSizer(hgap = 8, vgap = 2)
+            grid = wx.GridBagSizer(hgap = getSpace('minor'), vgap = getSpace('mini'))
             current_row = 0
+            # This gives the first entry on the page a tiny bit of extra space to the top
+            grid.Add((12, getSpace('mini')), pos = (current_row, 0), span = (1, 2))
+            current_row += 1
             for setting in settings:
                 if setting.name.startswith("separator_"):
                     sep = wx.StaticLine(grouppanel, size = (-1, 5), style = wx.LI_HORIZONTAL)
                     grid.Add(sep, pos = (current_row, 0), span = (1, 2),
-                             border = 3, flag = wx.ALIGN_CENTER | wx.ALL | wx.EXPAND)
+                             border = getSpace('mini'), flag = wx.ALIGN_CENTER | wx.ALL | wx.EXPAND)
                     current_row += 1
                 label, widget = setting.get_label(grouppanel), setting.get_widget(grouppanel)
                 if setting.name.startswith("separator_"):
                     font = label.GetFont()
                     font.SetWeight(wx.BOLD)
                     label.SetFont(font)
-                grid.Add(label, pos = (current_row, 0),
-                         flag = wx.ALIGN_CENTER_VERTICAL | wx.ALIGN_RIGHT)
+                grid.Add(label, pos = (current_row, 0), border = getSpace('minor'),
+                         flag = wx.ALIGN_CENTER_VERTICAL | wx.ALIGN_RIGHT | wx.LEFT)
                 expand = 0 if isinstance(widget, (wx.SpinCtrlDouble, wx.Choice, wx.ComboBox)) else wx.EXPAND
-                grid.Add(widget, pos = (current_row, 1),
-                         flag = wx.ALIGN_CENTER_VERTICAL | expand)
+                grid.Add(widget, pos = (current_row, 1), border = getSpace('minor'),
+                         flag = wx.ALIGN_CENTER_VERTICAL | wx.RIGHT | expand)
                 if hasattr(label, "set_default"):
                     label.Bind(wx.EVT_MOUSE_EVENTS, label.set_default)
                     if hasattr(widget, "Bind"):
@@ -171,13 +234,16 @@ class PronterOptionsDialog(wx.Dialog):
                 current_row += 1
             grid.AddGrowableCol(1)
             grouppanel.SetSizer(grid)
-        sbox.Add(notebook, 1, wx.EXPAND)
-        panel.SetSizer(sbox)
+            # The size of the options dialog is determined by the first panel 'Printer settings'
+            if group == group_list[0]:
+                grouppanel.SetMinSize(grid.ComputeFittingWindowSize(grouppanel))
         topsizer = wx.BoxSizer(wx.VERTICAL)
-        topsizer.Add(panel, 1, wx.ALL | wx.EXPAND)
-        topsizer.Add(self.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_CENTER)
-        self.SetSizerAndFit(topsizer)
-        self.SetMinSize(self.GetSize())
+        topsizer.Add(notebook, 1, wx.EXPAND | wx.ALL, getSpace('minor'))
+        topsizer.Add(wx.StaticLine(self, -1, style = wx.LI_HORIZONTAL), 0, wx.EXPAND)
+        topsizer.Add(self.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('minor'))
+        self.SetSizer(topsizer)
+        self.Fit()
+        self.CentreOnParent()
 
 notebookSelection = 0
 def PronterOptions(pronterface):
@@ -200,30 +266,61 @@ class ButtonEdit(wx.Dialog):
     """Custom button edit dialog"""
     def __init__(self, pronterface):
         wx.Dialog.__init__(self, None, title = _("Custom button"),
-                           style = wx.DEFAULT_DIALOG_STYLE | wx.RESIZE_BORDER)
+                           style = wx.DEFAULT_DIALOG_STYLE)
         self.pronterface = pronterface
-        topsizer = wx.BoxSizer(wx.VERTICAL)
-        grid = wx.FlexGridSizer(rows = 0, cols = 2, hgap = 4, vgap = 2)
+        # All widgets except ok/cancel are grouped into a static box for aesthetic and ux reasons
+        panel = wx.Panel(self)
+        #panelsizer = wx.BoxSizer
+        #self.sbox = wx.StaticBox(self)
+        #sboxsizer = wx.StaticBoxSizer(self.sbox, wx.VERTICAL)
+        grid = wx.FlexGridSizer(rows = 0, cols = 2, hgap = getSpace('minor'), vgap = getSpace('minor'))
         grid.AddGrowableCol(1, 1)
-        grid.Add(wx.StaticText(self, -1, _("Button title")), 0, wx.BOTTOM | wx.RIGHT)
-        self.name = wx.TextCtrl(self, -1, "")
+        ## Title of the button
+        grid.Add(wx.StaticText(panel, -1, _("Button title:")), 0, wx.ALIGN_CENTER_VERTICAL | wx.ALIGN_RIGHT)
+        self.name = wx.TextCtrl(panel, -1, "")
+        min_size = self.name.GetTextExtent('Default Long Button Name')
+        min_size.width = 260 #overwrite width to make the dialog a bit wider
+        self.name.SetMinSize(wx.Size(min_size.width, -1))
         grid.Add(self.name, 1, wx.EXPAND)
-        grid.Add(wx.StaticText(self, -1, _("Command")), 0, wx.BOTTOM | wx.RIGHT)
-        self.command = wx.TextCtrl(self, -1, "")
-        xbox = wx.BoxSizer(wx.HORIZONTAL)
-        xbox.Add(self.command, 1, wx.EXPAND)
+        ## Colour of the button
+        grid.Add(wx.StaticText(panel, -1, _("Colour:")), 0, wx.ALIGN_CENTER_VERTICAL | wx.ALIGN_RIGHT)
+        coloursizer = wx.BoxSizer(wx.HORIZONTAL)
+        self.use_colour = wx.CheckBox(panel,-1)
+        self.color = wx.ColourPickerCtrl(panel, colour=( 255, 255, 255), style=wx.CLRP_USE_TEXTCTRL)
+        self.color.Disable()
+        self.use_colour.Bind(wx.EVT_CHECKBOX, self.onColourCheckbox)
+        coloursizer.Add(self.use_colour,0, wx.ALIGN_CENTER_VERTICAL | wx.RIGHT, getSpace('minor'))
+        coloursizer.Add(self.color, 1, wx.EXPAND)
+        grid.Add(coloursizer,1, wx.EXPAND)
+        ## Enter commands or choose a macro
+        macrotooltip = _("Type short commands directly, enter a name for a new macro or select an existing macro from the list.")
+        commandfield = wx.StaticText(panel, -1, _("Command:"))
+        commandfield.SetToolTip(wx.ToolTip(macrotooltip))
+        grid.Add(commandfield, 0, wx.ALIGN_CENTER_VERTICAL | wx.ALIGN_RIGHT)
+        macrochoices = list(self.pronterface.macros.keys()) #add the available macros to the dropdown
+        macrochoices.insert(0,"") #add an empty field, so that a new macro name can be entered
+        self.command = wx.ComboBox(panel, -1, "", choices = macrochoices, style = wx.CB_DROPDOWN)
+        self.command.SetToolTip(wx.ToolTip(macrotooltip))
+        commandsizer = wx.BoxSizer(wx.HORIZONTAL)
+        commandsizer.Add(self.command, 1, wx.EXPAND)
         self.command.Bind(wx.EVT_TEXT, self.macrob_enabler)
-        self.macrob = wx.Button(self, -1, "..", style = wx.BU_EXACTFIT)
-        self.macrob.Bind(wx.EVT_BUTTON, self.macrob_handler)
-        xbox.Add(self.macrob, 0)
-        grid.Add(xbox, 1, wx.EXPAND)
-        grid.Add(wx.StaticText(self, -1, _("Color")), 0, wx.BOTTOM | wx.RIGHT)
-        self.color = wx.TextCtrl(self, -1, "")
-        grid.Add(self.color, 1, wx.EXPAND)
-        topsizer.Add(grid, 0, wx.EXPAND)
-        topsizer.Add((0, 0), 1)
-        topsizer.Add(self.CreateStdDialogButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_CENTER)
+        self.macrobtn = wx.Button(panel, -1, "...", style = wx.BU_EXACTFIT)
+        self.macrobtn.SetMinSize((self.macrobtn.GetTextExtent('AAA').width, -1))
+        self.macrobtn.SetToolTip(wx.ToolTip(_("Create a new macro or edit an existing one.")))
+        self.macrobtn.Bind(wx.EVT_BUTTON, self.macrob_handler)
+        commandsizer.Add(self.macrobtn, 0, wx.ALIGN_CENTER_VERTICAL | wx.LEFT, getSpace('minor'))
+        grid.Add(commandsizer, 1, wx.EXPAND)
+        ## Put it in a box and add ok/cancel buttons
+        #panel.Add(grid, 1, wx.EXPAND | wx.ALL, getSpace('mini'))
+        panel.SetSizer(grid)
+        topsizer = wx.BoxSizer(wx.VERTICAL)
+        topsizer.Add(panel, 0, wx.EXPAND | wx.ALL, getSpace('major'))
+        topsizer.Add(wx.StaticLine(self, -1, style = wx.LI_HORIZONTAL), 0, wx.EXPAND)
+        topsizer.Add(self.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('minor'))
         self.SetSizer(topsizer)
+        topsizer.Fit(self)
+        self.CentreOnParent()
+        self.name.SetFocus()
 
     def macrob_enabler(self, e):
         macro = self.command.GetValue()
@@ -248,7 +345,7 @@ class ButtonEdit(wx.Dialog):
                 valid = False
             else:
                 valid = True
-        self.macrob.Enable(valid)
+        self.macrobtn.Enable(valid)
 
     def macrob_handler(self, e):
         macro = self.command.GetValue()
@@ -256,6 +353,13 @@ class ButtonEdit(wx.Dialog):
         self.command.SetValue(macro)
         if self.name.GetValue() == "":
             self.name.SetValue(macro)
+
+    def onColourCheckbox(self, e):
+        status = self.use_colour.GetValue()
+        if status:
+            self.color.Enable()
+        else:
+            self.color.Disable()
 
 class TempGauge(wx.Panel):
 

--- a/printrun/pronterface.py
+++ b/printrun/pronterface.py
@@ -49,7 +49,7 @@ except:
     logging.error(_("WX >= 4 is not installed. This program requires WX >= 4 to run."))
     raise
 
-from .gui.widgets import SpecialButton, MacroEditor, PronterOptions, ButtonEdit
+from .gui.widgets import SpecialButton, MacroEditor, PronterOptions, ButtonEdit, getSpace
 
 winsize = (800, 500)
 layerindex = 0
@@ -1077,6 +1077,7 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
         self.settings._add(BooleanSetting("monitor", True, _("Monitor printer status"), _("Regularly monitor printer temperatures (required to have functional temperature graph or gauges)"), "Printer"), self.update_monitor)
         self.settings._add(StringSetting("simarrange_path", "", _("Simarrange command"), _("Path to the simarrange binary to use in the STL plater"), "External"))
         self.settings._add(BooleanSetting("circular_bed", False, _("Circular build platform"), _("Draw a circular (or oval) build platform instead of a rectangular one"), "Printer"), self.update_bed_viz)
+        #self.settings._add(ComboSetting("extruders", "1", ("0", "1", "2", "3", "4", "5"), _("Extruders count"), _("Number of extruders"), "Printer"))
         self.settings._add(SpinSetting("extruders", 0, 1, 5, _("Extruders count"), _("Number of extruders"), "Printer"))
         self.settings._add(BooleanSetting("clamp_jogging", False, _("Clamp manual moves"), _("Prevent manual moves from leaving the specified build dimensions"), "Printer"))
         self.settings._add(BooleanSetting("display_progress_on_printer", False, _("Display progress on printer"), _("Show progress on printers display (sent via M117, might not be supported by all printers)"), "Printer"))
@@ -1084,15 +1085,15 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
         self.settings._add(BooleanSetting("cutting_as_extrusion", True, _("Display cutting moves"), _("Show moves where spindle is active as printing moves"), "Printer"))
         self.settings._add(ComboSetting("uimode", _("Standard"), [_("Standard"), _("Compact"), ], _("Interface mode"), _("Standard interface is a one-page, three columns layout with controls/visualization/log\nCompact mode is a one-page, two columns layout with controls + log/visualization"), "UI"), self.reload_ui)
         #self.settings._add(ComboSetting("uimode", _("Standard"), [_("Standard"), _("Compact"), _("Tabbed"), _("Tabbed with platers")], _("Interface mode"), _("Standard interface is a one-page, three columns layout with controls/visualization/log\nCompact mode is a one-page, two columns layout with controls + log/visualization"), "UI"), self.reload_ui)
-        self.settings._add(ComboSetting("controlsmode", "Standard", ("Standard", "Mini"), _("Controls mode"), _("Standard controls include all controls needed for printer setup and calibration, while Mini controls are limited to the ones needed for daily printing"), "UI"), self.reload_ui)
+        self.settings._add(ComboSetting("controlsmode", _("Standard"), (_("Standard"), _("Mini"), ), _("Controls mode"), _("Standard controls include all controls needed for printer setup and calibration, while Mini controls are limited to the ones needed for daily printing"), "UI"), self.reload_ui)
         self.settings._add(BooleanSetting("slic3rintegration", False, _("Enable Slic3r integration"), _("Add a menu to select Slic3r profiles directly from Pronterface"), "UI"), self.reload_ui)
         self.settings._add(BooleanSetting("slic3rupdate", False, _("Update Slic3r default presets"), _("When selecting a profile in Slic3r integration menu, also save it as the default Slic3r preset"), "UI"))
-        self.settings._add(ComboSetting("mainviz", "3D", ("2D", "3D", "None"), _("Main visualization"), _("Select visualization for main window."), "Viewer"), self.reload_ui)
+        self.settings._add(ComboSetting("mainviz", "3D", ("2D", "3D", _("None")), _("Main visualization"), _("Select visualization for main window."), "Viewer"), self.reload_ui)
         self.settings._add(BooleanSetting("viz3d", False, _("Use 3D in GCode viewer window"), _("Use 3D mode instead of 2D layered mode in the visualization window"), "Viewer"), self.reload_ui)
         self.settings._add(StaticTextSetting("separator_3d_viewer", _("3D viewer options"), "", group = "Viewer"))
         self.settings._add(BooleanSetting("light3d", False, _("Use a lighter 3D visualization"), _("Use a lighter visualization with simple lines instead of extruded paths for 3D viewer"), "Viewer"), self.reload_ui)
         self.settings._add(BooleanSetting("perspective", False, _("Use a perspective view instead of orthographic"), _("A perspective view looks more realistic, but is a bit more confusing to navigate"), "Viewer"), self.reload_ui)
-        self.settings._add(ComboSetting("antialias3dsamples", "0", ["0", "2", "4", "8"], _("Number of anti-aliasing samples"), _("Amount of anti-aliasing samples used in the 3D viewer"), "Viewer"), self.reload_ui)
+        self.settings._add(ComboSetting("antialias3dsamples", "0", ("0", "2", "4", "8"), _("Number of anti-aliasing samples"), _("Amount of anti-aliasing samples used in the 3D viewer"), "Viewer"), self.reload_ui)
         self.settings._add(BooleanSetting("trackcurrentlayer3d", False, _("Track current layer in main 3D view"), _("Track the currently printing layer in the main 3D visualization"), "Viewer"))
         self.settings._add(FloatSpinSetting("gcview_path_width", 0.4, 0.01, 2, _("Extrusion width for 3D viewer"), _("Width of printed path in 3D viewer"), "Viewer", increment = 0.05), self.update_gcview_params)
         self.settings._add(FloatSpinSetting("gcview_path_height", 0.3, 0.01, 2, _("Layer height for 3D viewer"), _("Height of printed path in 3D viewer"), "Viewer", increment = 0.05), self.update_gcview_params)
@@ -1113,7 +1114,9 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
         self.settings._add(FloatSpinSetting("preview_extrusion_width", 0.5, 0, 10, _("Preview extrusion width"), _("Width of Extrusion in Preview"), "Viewer", increment = 0.1), self.update_gviz_params)
         self.settings._add(SpinSetting("preview_grid_step1", 10., 0, 200, _("Fine grid spacing"), _("Fine Grid Spacing"), "Viewer"), self.update_gviz_params)
         self.settings._add(SpinSetting("preview_grid_step2", 50., 0, 200, _("Coarse grid spacing"), _("Coarse Grid Spacing"), "Viewer"), self.update_gviz_params)
+        self.settings._add(StaticTextSetting("separator_colors1", _("General"), "", group = "Colors"))
         self.settings._add(ColorSetting("bgcolor", self._preferred_bgcolour_hex(), _("Background color"), _("Pronterface background color"), "Colors", isRGBA=False), self.reload_ui)
+        self.settings._add(StaticTextSetting("separator_colors2", _("Temperature Graph"), "", group = "Colors"))
         self.settings._add(ColorSetting("graph_color_background", "#FAFAC7", _("Graph background color"), _("Color of the temperature graph background"), "Colors", isRGBA=False), self.reload_ui)
         self.settings._add(ColorSetting("graph_color_text", "#172C2C", _("Graph text color"), _("Color of the temperature graph text"), "Colors", isRGBA=False), self.reload_ui)
         self.settings._add(ColorSetting("graph_color_grid", "#5A5A5A", _("Graph grid color"), _("Color of the temperature graph grid"), "Colors", isRGBA=False), self.reload_ui)
@@ -1124,6 +1127,7 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
         self.settings._add(ColorSetting("graph_color_ex0target", "#0005FF80", _("Graph ex0 target line color"), _("Color of the temperature graph extruder 0 target temperature line"), "Colors"), self.reload_ui)
         self.settings._add(ColorSetting("graph_color_ex1temp", "#37370080", _("Graph ex1 line color color"), _("Color of the temperature graph extruder 1 temperature line"), "Colors"), self.reload_ui)
         self.settings._add(ColorSetting("graph_color_ex1target", "#37370080", _("Graph ex1 target line color"), _("Color of the temperature graph extruder 1 temperature target line"), "Colors"), self.reload_ui)
+        self.settings._add(StaticTextSetting("separator_colors3", _("3D Viewer"), "", group = "Colors"))
         self.settings._add(ColorSetting("gcview_color_background", "#FAFAC7FF", _("3D view background color"), _("Color of the 3D view background"), "Colors"), self.update_gcview_colors)
         self.settings._add(ColorSetting("gcview_color_travel", "#99999999", _("3D view travel moves color"), _("Color of travel moves in 3D view"), "Colors"), self.update_gcview_colors)
         self.settings._add(ColorSetting("gcview_color_tool0", "#FF000099", _("3D view print moves color"), _("Color of print moves with tool 0 in 3D view"), "Colors"), self.update_gcview_colors)
@@ -2130,8 +2134,8 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
             if btndef is None:
                 if i == len(custombuttons) - 1:
                     self.newbuttonbutton = b = wx.Button(self.centerpanel, -1, "+", size = (35, 18), style = wx.BU_EXACTFIT)
-                    b.SetForegroundColour("#4444ff")
-                    b.SetToolTip(wx.ToolTip(_("click to add new custom button")))
+                    #b.SetForegroundColour("#4444ff")
+                    b.SetToolTip(wx.ToolTip(_("Click to add new custom button")))
                     b.Bind(wx.EVT_BUTTON, self.cbutton_edit)
                 else:
                     b = wx.StaticText(self.panel, -1, "")
@@ -2140,8 +2144,7 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
                 b.SetToolTip(wx.ToolTip(_("Execute command: ") + btndef.command))
                 if btndef.background:
                     b.SetBackgroundColour(btndef.background)
-                    rr, gg, bb, aa = b.GetBackgroundColour().Get() #last item is alpha
-                    if 0.3 * rr + 0.59 * gg + 0.11 * bb < 60:
+                    if b.GetBackgroundColour().GetLuminance() < 0.5:
                         b.SetForegroundColour("#ffffff")
                 b.custombutton = i
                 b.properties = btndef
@@ -2219,7 +2222,9 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
                         colour = wx.Colour(*colour).GetAsString(wx.C2S_NAME | wx.C2S_HTML_SYNTAX)
                     else:
                         colour = wx.Colour(colour).GetAsString(wx.C2S_NAME | wx.C2S_HTML_SYNTAX)
-                bedit.color.SetValue(colour)
+                bedit.use_colour.SetValue(True)
+                bedit.color.Enable()
+                bedit.color.SetColour(colour)
         else:
             n = len(self.custombuttons)
             while n > 0 and self.custombuttons[n - 1] is None:
@@ -2228,8 +2233,8 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
             if n == len(self.custombuttons):
                 self.custombuttons.append(None)
             self.custombuttons[n] = SpecialButton(bedit.name.GetValue().strip(), bedit.command.GetValue().strip(), custom = True)
-            if bedit.color.GetValue().strip() != "":
-                self.custombuttons[n].background = bedit.color.GetValue()
+            if bedit.use_colour.GetValue():
+                self.custombuttons[n].background = bedit.color.GetColour().GetAsString(wx.C2S_CSS_SYNTAX)
             self.cbutton_save(n, self.custombuttons[n])
         wx.CallAfter(bedit.Destroy)
         wx.CallAfter(self.cbuttons_reload)
@@ -2271,6 +2276,8 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
                 item = popupmenu.Append(-1, _("Move right >>"))
                 self.Bind(wx.EVT_MENU, lambda e, button = e.GetEventObject(): self.cbutton_order(e, button, 1), item)
                 if obj.custombutton == 63: item.Enable(False)
+                buttonscount = len(getattr(self, "custombuttons_widgets", [])) - 2
+                if obj.custombutton == buttonscount: item.Enable(False)
                 pos = self.panel.ScreenToClient(e.GetEventObject().ClientToScreen(pos))
                 item = popupmenu.Append(-1, _("Remove custom button '%s'") % e.GetEventObject().GetLabelText())
                 self.Bind(wx.EVT_MENU, lambda e, button = e.GetEventObject(): self.cbutton_remove(e, button), item)
@@ -2419,34 +2426,30 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
 
     def new_macro(self, e = None):
         dialog = wx.Dialog(self, -1, _("Enter macro name"))
-        text = wx.StaticText(dialog, -1, _("Macro name:"))
-        namectrl = wx.TextCtrl(dialog, -1, style = wx.TE_PROCESS_ENTER)
-        okb = wx.Button(dialog, wx.ID_OK, _("Ok"))
+        panel = wx.Panel(dialog)
+        textsizer = wx.BoxSizer(wx.HORIZONTAL)
+        text = wx.StaticText(panel, -1, _("Macro name:"))
+        namectrl = wx.TextCtrl(panel, -1, style = wx.TE_PROCESS_ENTER)
         dialog.Bind(wx.EVT_TEXT_ENTER,
             lambda e: dialog.EndModal(wx.ID_OK), namectrl)
-        cancel_button = wx.Button(dialog, wx.ID_CANCEL, _("Cancel"))
-
         # Layout
-        ## Group the buttons horizontally
-        buttons_sizer = wx.BoxSizer(wx.HORIZONTAL)
-        buttons_sizer.Add(okb, 0)
-        buttons_sizer.Add(cancel_button, 0)
         ## Set a minimum size for the name control box
         min_size = namectrl.GetTextExtent('Default Long Macro Name')
         namectrl.SetMinSize(wx.Size(min_size.width, -1))
         ## Group the text and the name control box horizontally
-        name_sizer = wx.BoxSizer(wx.HORIZONTAL)
-        name_sizer.Add(text, 0, flag = wx.ALIGN_CENTER)
-        name_sizer.AddSpacer(10)
-        name_sizer.Add(namectrl, 1, wx.EXPAND)
+        topsizer = wx.BoxSizer(wx.VERTICAL)
+        textsizer.Add(text, 0, wx.ALIGN_CENTER_VERTICAL | wx.ALIGN_LEFT)
+        textsizer.AddSpacer(getSpace('minor'))
+        textsizer.Add(namectrl, 1, wx.EXPAND | wx.ALIGN_LEFT)
+        panel.SetSizer(textsizer)
+        topsizer.Add(panel,1, wx.ALL, getSpace('major'))
         ## Group everything vertically
-        dialog_sizer = wx.BoxSizer(wx.VERTICAL)
-        dialog_sizer.Add(name_sizer, 0, border = 10,
-            flag = wx.LEFT | wx.TOP | wx.RIGHT)
-        dialog_sizer.Add(buttons_sizer, 0, border = 10,
-            flag = wx.ALIGN_CENTER | wx.ALL)
-        dialog.SetSizerAndFit(dialog_sizer)
-        dialog.Centre()
+        topsizer.Add(wx.StaticLine(dialog, -1, style = wx.LI_HORIZONTAL), 0, wx.EXPAND)
+        topsizer.Add(dialog.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('minor'))
+        dialog.SetSizer(topsizer)
+        topsizer.Fit(dialog)
+        dialog.CentreOnParent()
+        namectrl.SetFocus()
 
         macro = ""
         if dialog.ShowModal() == wx.ID_OK:

--- a/printrun/pronterface.py
+++ b/printrun/pronterface.py
@@ -1077,7 +1077,6 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
         self.settings._add(BooleanSetting("monitor", True, _("Monitor printer status"), _("Regularly monitor printer temperatures (required to have functional temperature graph or gauges)"), "Printer"), self.update_monitor)
         self.settings._add(StringSetting("simarrange_path", "", _("Simarrange command"), _("Path to the simarrange binary to use in the STL plater"), "External"))
         self.settings._add(BooleanSetting("circular_bed", False, _("Circular build platform"), _("Draw a circular (or oval) build platform instead of a rectangular one"), "Printer"), self.update_bed_viz)
-        #self.settings._add(ComboSetting("extruders", "1", ("0", "1", "2", "3", "4", "5"), _("Extruders count"), _("Number of extruders"), "Printer"))
         self.settings._add(SpinSetting("extruders", 0, 1, 5, _("Extruders count"), _("Number of extruders"), "Printer"))
         self.settings._add(BooleanSetting("clamp_jogging", False, _("Clamp manual moves"), _("Prevent manual moves from leaving the specified build dimensions"), "Printer"))
         self.settings._add(BooleanSetting("display_progress_on_printer", False, _("Display progress on printer"), _("Show progress on printers display (sent via M117, might not be supported by all printers)"), "Printer"))
@@ -2134,7 +2133,6 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
             if btndef is None:
                 if i == len(custombuttons) - 1:
                     self.newbuttonbutton = b = wx.Button(self.centerpanel, -1, "+", size = (35, 18), style = wx.BU_EXACTFIT)
-                    #b.SetForegroundColour("#4444ff")
                     b.SetToolTip(wx.ToolTip(_("Click to add new custom button")))
                     b.Bind(wx.EVT_BUTTON, self.cbutton_edit)
                 else:
@@ -2276,7 +2274,7 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
                 item = popupmenu.Append(-1, _("Move right >>"))
                 self.Bind(wx.EVT_MENU, lambda e, button = e.GetEventObject(): self.cbutton_order(e, button, 1), item)
                 if obj.custombutton == 63: item.Enable(False)
-                buttonscount = len(getattr(self, "custombuttons_widgets", [])) - 2
+                buttonscount = len(self.custombuttons_widgets) - 2
                 if obj.custombutton == buttonscount: item.Enable(False)
                 pos = self.panel.ScreenToClient(e.GetEventObject().ClientToScreen(pos))
                 item = popupmenu.Append(-1, _("Remove custom button '%s'") % e.GetEventObject().GetLabelText())
@@ -2445,7 +2443,7 @@ Printrun. If not, see <http://www.gnu.org/licenses/>."""
         topsizer.Add(panel,1, wx.ALL, getSpace('major'))
         ## Group everything vertically
         topsizer.Add(wx.StaticLine(dialog, -1, style = wx.LI_HORIZONTAL), 0, wx.EXPAND)
-        topsizer.Add(dialog.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('minor'))
+        topsizer.Add(dialog.CreateButtonSizer(wx.OK | wx.CANCEL), 0, wx.ALIGN_RIGHT | wx.ALL, getSpace('stddlg'))
         dialog.SetSizer(topsizer)
         topsizer.Fit(dialog)
         dialog.CentreOnParent()

